yara-x/src/modules/macho/parser.rs

```diff
@@ -0,0 +1,1079 @@
+        )
+    }
+
+    fn sparc_thread_state(
+        &self,
+    ) -> impl FnMut(&'a [u8]) -> IResult<&'a [u8], u64> + '_ {
+        map(
+            tuple((
+                u32(self.endianness),           // psr
+                u32(self.endianness),           // pc
+                u32(self.endianness),           // npc
+                u32(self.endianness),           // y
+                count(u32(self.endianness), 7), // g
+                count(u32(self.endianness), 7), // o
+            )),
+            |(_, pc, _, _, _, _)| pc as u64,
+        )
+    }
+
+    fn m68k_thread_state(
+        &self,
+    ) -> impl FnMut(&'a [u8]) -> IResult<&'a [u8], u64> + '_ {
+        map(
+            tuple((
+                count(u32(self.endianness), 8), // dreg
+                count(u32(self.endianness), 8), // areg
+                u16(self.endianness),           // pad
+                u16(self.endianness),           // sr
+                u32(self.endianness),           // pc
+            )),
+            |(_, _, _, _, pc)| pc as u64,
+        )
+    }
+
+    fn m88k_thread_state(
+        &self,
+    ) -> impl FnMut(&'a [u8]) -> IResult<&'a [u8], u64> + '_ {
+        map(
+            tuple((
+                count(u32(self.endianness), 31), // r
+                u32(self.endianness),            // xip
+                u32(self.endianness),            // xip_in_bd
+                u32(self.endianness),            // nip
+            )),
+            |(_, xip, _, _)| xip as u64,
+        )
+    }
+}
+
+struct FatArch {
+    cputype: u32,
+    cpusubtype: u32,
+    offset: u64,
+    size: u64,
+    align: u32,
+    reserved: u32,
+}
+
+struct MachOHeader {
+    magic: u32,
+    cputype: u32,
+    cpusubtype: u32,
+    filetype: u32,
+    ncmds: u32,
+    sizeofcmds: u32,
+    flags: u32,
+    reserved: Option<u32>, // Only set in 64-bits binary.
+}
+
+struct Segment<'a> {
+    segname: &'a [u8],
+    vmaddr: u64,
+    vmsize: u64,
+    fileoff: u64,
+    filesize: u64,
+    maxprot: u32,
+    initprot: u32,
+    nsects: u32,
+    flags: u32,
+    sections: Vec<Section<'a>>,
+}
+
+struct Section<'a> {
+    sectname: &'a [u8],
+    segname: &'a [u8],
+    addr: u64,
+    size: u64,
+    offset: u32,
+    align: u32,
+    reloff: u32,
+    nreloc: u32,
+    flags: u32,
+    reserved1: u32,
+    reserved2: u32,
+    reserved3: Option<u32>, // Only set in 64-bits binaries
+}
+
+struct Dylib<'a> {
+    name: &'a [u8],
+    timestamp: u32,
+    current_version: u32,
+    compatibility_version: u32,
+}
+
+struct Dysymtab {
+    ilocalsym: u32,
+    nlocalsym: u32,
+    iextdefsym: u32,
+    nextdefsym: u32,
+    tocoff: u32,
+    ntoc: u32,
+    modtaboff: u32,
+    nmodtab: u32,
+    extrefsymoff: u32,
+    nextrefsyms: u32,
+    indirectsymoff: u32,
+    nindirectsyms: u32,
+    extreloff: u32,
+    nextrel: u32,
+    locreloff: u32,
+    nlocrel: u32,
+}
+
+/// Parser that reads a 32-bits or 64-bits
+fn uint(
```

Comment:
we do have some fields in certain mach-o structs that will be u8 sizes, do we want to account for that here?

example: https://opensource.apple.com/source/xnu/xnu-4570.61.1/osfmk/kern/cs_blobs.h.auto.html

```c
typedef struct __CodeDirectory {
	uint32_t magic;					/* magic number (CSMAGIC_CODEDIRECTORY) */
	uint32_t length;				/* total length of CodeDirectory blob */
	uint32_t version;				/* compatibility version */
	uint32_t flags;					/* setup and mode flags */
	uint32_t hashOffset;			/* offset of hash slot element at index zero */
	uint32_t identOffset;			/* offset of identifier string */
	uint32_t nSpecialSlots;			/* number of special hash slots */
	uint32_t nCodeSlots;			/* number of ordinary (code) hash slots */
	uint32_t codeLimit;				/* limit to main image signature range */
	uint8_t hashSize;				/* size of each hash in bytes */
	uint8_t hashType;				/* type of hash (cdHashType* constants) */
	uint8_t platform;				/* platform identifier; zero if not platform binary */
	uint8_t	pageSize;				/* log2(page size in bytes); 0 => infinite */
	uint32_t spare2;				/* unused (must be zero) */
}
```

Reply:
This `uint` parser is intended to be used in cases where the integer's endianness, or wideness (or both) is not known at compile time. `u8` doesn't have those problems as they are not affected by endianness and the size if already known.

Reply:
ah ok great, sounds good to me. We do have some instances where endianness is always big for parsing certain mach-o structs as well, but that is also easily handled already with nom. thanks for the sanity check!